core/sessions/oidcauth/oidc.go

```diff
@@ -0,0 +1,675 @@
+		return clsessions.ErrNotSupported
+	}
+
+	// User is local admin, save new password
+	hashedPassword, err := utils.HashPassword(newPassword)
+	if err != nil {
+		return err
+	}
+	if err := sqlutil.TransactDataSource(ctx, oi.ds, nil, func(tx sqlutil.DataSource) error {
+		sql := "UPDATE users SET hashed_password = $1, updated_at = now() WHERE email = $2 RETURNING *"
+		return tx.GetContext(ctx, user, sql, hashedPassword, user.Email)
+	}); err != nil {
+		oi.lggr.Errorf("unable to set password for user: %s, err: %v", user.Email, err)
+		return errors.New("unable to save password")
+	}
+	return nil
+}
+
+// TestPassword only supports the potential local admin user, as there is no queryable identity server for the OIDC implementation
+func (oi *oidcAuthenticator) TestPassword(ctx context.Context, email string, password string) error {
+	// Fall back to test local users table in case of supported local CLI users as well
+	var hashedPassword string
+	if err := oi.ds.GetContext(ctx, &hashedPassword, "SELECT hashed_password FROM users WHERE lower(email) = lower($1)", email); err != nil {
+		return errors.New("invalid credentials")
+	}
+	if !utils.CheckPasswordHash(password, hashedPassword) {
+		return errors.New("invalid credentials")
+	}
+	return nil
+}
+
+// CreateAndSetAuthToken generates a new credential token with the user role
+func (oi *oidcAuthenticator) CreateAndSetAuthToken(ctx context.Context, user *clsessions.User) (*auth.Token, error) {
+	newToken := auth.NewToken()
+	err := oi.SetAuthToken(ctx, user, newToken)
+	if err != nil {
+		return nil, err
+	}
+	return newToken, nil
+}
+
+// SetAuthToken updates the user to use the given Authentication Token.
+func (oi *oidcAuthenticator) SetAuthToken(ctx context.Context, user *clsessions.User, token *auth.Token) error {
+	if !oi.config.UserAPITokenEnabled() {
+		return errors.New("API token is not enabled ")
+	}
+
+	salt := utils.NewSecret(utils.DefaultSecretSize)
+	hashedSecret, err := auth.HashedSecret(token, salt)
+	if err != nil {
+		return fmt.Errorf("OIDCAuth SetAuthToken hashed secret error: %w", err)
+	}
+
+	err = sqlutil.TransactDataSource(ctx, oi.ds, nil, func(tx sqlutil.DataSource) error {
+		// Remove any existing API tokens
+		if _, err = oi.ds.ExecContext(ctx, "DELETE FROM oidc_user_api_tokens WHERE user_email = $1", user.Email); err != nil {
+			return fmt.Errorf("error executing DELETE FROM oidc_user_api_tokens: %w", err)
+		}
+		// Create new API token for user
+		_, err = oi.ds.ExecContext(ctx,
+			"INSERT INTO oidc_user_api_tokens (user_email, user_role, token_key, token_salt, token_hashed_secret, created_at) VALUES ($1, $2, $3, $4, $5, $6, now())",
+			user.Email,
+			user.Role,
+			token.AccessKey,
+			salt,
+			hashedSecret,
+		)
+		if err != nil {
+			return fmt.Errorf("failed insert into oidc_user_api_tokens: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.New("error creating API token")
```

Comment:
Should `err` be wrapped? Or logged?
```suggestion
		return fmt.Errorf("error creating API token: %w", err)
```

Reply:
Yes we should definitely log, looks like this error is returned to the user so added a logger error call